org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/RecyclableObjectPool.java

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.debug.adapter;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * An utility object pool class with the following  ability: 
+ * 1. store an object to get an object id.
+ * 2. remove an object.
+ * 3. remove objects which has specified owner.
+ * 4. remove all objects.
+ *
+ * <p>It is thread-safe, the duplicate object will not be stored, an object can be referenced by multiple owners, it is
+ * removed only when user explicitly calls removeObjectById or all the owners has been removed.</p>
+ *
+ * @param <O> the owner class type
+ * @param <V> the object type
+ */
+public class RecyclableObjectPool<O, V> {
```

Comment:
The class is not used for recycling IDs, but recycling un-referenced  JDI objects to save memory, orelse all the history vairables will be saved in objectCollection which will consume more and more memory which leads to OOM.